api/v1alpha1/dataplane_konnect_extension_types.go

```diff
@@ -0,0 +1,98 @@
+package v1alpha1
+
+/*
+Copyright 2024 Kong Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+func init() {
+	SchemeBuilder.Register(&DataPlaneKonnectExtension{}, &DataPlaneKonnectExtensionList{})
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=kong;all
+// +kubebuilder:subresource:status
+
+// DataPlaneKonnectExtension is the Schema for the dataplanekonnectextension API,
+// and is intended to be referenced as extension by the dataplane API.
+// If a DataPlane successfully refers a DataPlaneKonnectExtension, the DataPlane
+// deployment spec gets customized to include the konnect-related configuration.
+type DataPlaneKonnectExtension struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the specification of the DataPlaneKonnectExtension resource.
+	Spec DataPlaneKonnectExtensionSpec `json:"spec,omitempty"`
+	// Status is the status of the DataPlaneKonnectExtension resource.
+	Status DataPlaneKonnectExtensionStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// DataPlaneKonnectExtensionList contains a list of DataPlaneKonnectExtension.
+type DataPlaneKonnectExtensionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DataPlaneKonnectExtension `json:"items"`
+}
+
+// DataPlaneKonnectExtensionSpec defines the desired state of DataPlaneKonnectExtension.
+type DataPlaneKonnectExtensionSpec struct {
+	// ControlPlaneRef is a reference to a ControlPlane this DataPlaneKonnectExtension is associated with.
+	// +kubebuilder:validation:Required
+	ControlPlaneRef configurationv1alpha1.ControlPlaneRef `json:"controlPlaneRef"`
```

Comment:
Just a note based on a sync with @mlavacca: when we get to implement the konnectNamespacedRef we should make the CP ref immutable to prevent going from 1 type of CP ref to another (unless that gets implemented)